# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmBCClass.h
# Commit: 40f861f62a0f
# Full Hash: 40f861f62a0f58040cb6cadf073bc5e97a44231d
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2021-12-23 09:34:36
# Regressor Bug: 1642412
# File Overlap Count: 7
# Description:
#   Bug 1642412 - wasm: Optimize baseline table.size/get/set. r=lth
#   
#   This commit adds a fast path for table.get/set when the table
#   representation is anyref, and an unconditional fast path for
#   table.length.
# ==============================================================================

diff -r 58a45475b524 -r 40f861f62a0f js/src/wasm/WasmBCClass.h
--- a/js/src/wasm/WasmBCClass.h	Wed Dec 22 19:09:54 2021 +0200
+++ b/js/src/wasm/WasmBCClass.h	Wed Dec 22 17:36:29 2021 +0000
@@ -1070,6 +1070,15 @@
 
   //////////////////////////////////////////////////////////////////////
   //
+  // Table access.
+
+  Address addressOfTableField(const TableDesc& table, uint32_t fieldOffset,
+                              RegPtr tls);
+  void loadTableLength(const TableDesc& table, RegPtr tls, RegI32 length);
+  void loadTableElements(const TableDesc& table, RegPtr tls, RegPtr elements);
+
+  //////////////////////////////////////////////////////////////////////
+  //
   // Heap access.
 
   void bceCheckLocal(MemoryAccessDesc* access, AccessCheck* check,
@@ -1525,6 +1534,10 @@
   [[nodiscard]] bool emitTableSet();
   [[nodiscard]] bool emitTableSize();
 
+  void emitTableBoundsCheck(const TableDesc& table, RegI32 index, RegPtr tls);
+  [[nodiscard]] bool emitTableGetAnyRef(uint32_t tableIndex);
+  [[nodiscard]] bool emitTableSetAnyRef(uint32_t tableIndex);
+
 #ifdef ENABLE_WASM_GC
   [[nodiscard]] bool emitStructNewWithRtt();
   [[nodiscard]] bool emitStructNewDefaultWithRtt();